### PR TITLE
[core] Bind jemalloc to raylet and GCS

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
The default malloc failed to do a good job in managing the memory. For example, the memory fragmentation is high after running the server for a while. jemalloc mitigated this issue.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -113,6 +114,7 @@ build:asan --copt -fsanitize=address
 build:asan --copt -DADDRESS_SANITIZER
 build:asan --copt -fno-omit-frame-pointer
 build:asan --linkopt -fsanitize=address
+build:asan --no//:jemalloc_flag
```
Comment (rkooo567) [inline .bazelrc:117]: Q: why not?

Comment (rkooo567) [reply .bazelrc:117]: can you maybe comment it to the code?

File: BUILD.bazel [hunk 2]
```diff
@@ -74,8 +75,31 @@ config_setting(
...
-# === Begin of rpc definitions ===
+bool_flag(
+    name = "jemalloc_flag",
+    build_setting_default = True,
+)
+
+config_setting(
+    name = "jemalloc",
+    constraint_values = [
+        "@platforms//os:linux",
```
Comment (rkooo567) [inline BUILD.bazel:86]: Does it mean if this constraints are not satisfied, it won't be built with jemalloc?

**Review by rkooo567 — APPROVED** [review]
Nice. How can we verify if our ray is built with jemalloc properly btw?

File: .bazelrc (hunk 1 above)
Comment (iycheng) [reply .bazelrc:117]: Asan test will inject their own malloc and with jemalloc, it'll crash :(

File: BUILD.bazel (hunk 2 above)
Comment (iycheng) [reply BUILD.bazel:86]: Yes. Right now I only enable it for linux on x86.

**Comment by iycheng** [comment]
I'll add the comment later given the cut is 5pm today.

**Comment by iycheng** [comment]
> Nice. How can we verify if our ray is built with jemalloc properly btw?
it's easy:
```
$ nm bazel-bin/gcs_server | grep jemalloc
0000000000cb1db0 t je_jemalloc_postfork_child
0000000000cb1b80 t je_jemalloc_postfork_parent
0000000000cb19e0 t je_jemalloc_prefork
000000000020fea0 t jemalloc_constructor
```

**Pull request merged**